### DISCUSSION TITLE
How to calculate how much resources a SQL uses, including CPU and memory

### DISCUSSION THREAD

**Discussion by A-little-bit-of-data** [trinodb/trino#24407] [Q&A]
The cluster has one coordinator and seven workers. The coordinator configuration is 16C 64G, and the worker configuration is 12C 64G.
I ran a SQL statement and the following information was displayed on the web UI:

```
Elapsed Time	58.70s
Queued Time	39.28us
Analysis Time	767.47ms
Planning Time	150.23ms
Execution Time	57.93s
CPU Time	4.21s
Planning CPU Time	66.98ms
Scheduled Time	10.73m
Input Rows	515K
Input Data	22.4MB
Physical Input Rows	515K
Physical Input Data	0B
Physical Input Read Time	18.01s
Internal Network Rows	784K
Internal Network Data	56.4MB
Peak User Memory	78.4MB
Peak Total Memory	78.4MB
Cumulative User Memory	1.94GB*seconds
Output Rows	1.00
Output Data	9.00B
Written Rows	53.6K
Logical Written Data	10.6MB
Physical Written Data	1.15MB
```

How to calculate how much resources this SQL task consumes?

**Comment by hashhar** [comment]
it's right there in your output. 78.4MB of memory during peak consumption - 1.94GB is the integral of memory usage over the query duration (hence GB-seconds). 4.21s of CPU time spread out across the cluster of which 66.98ms was spent during planning on the coordinator.

However unless you define what you want to do the numbers are useless.

**Reply by A-little-bit-of-data** [reply]
> it's right there in your output. 78.4MB of memory during peak consumption - 1.94GB is the integral of memory usage over the query duration (hence GB-seconds). 4.21s of CPU time spread out across the cluster of which 66.98ms was spent during planning on the coordinator.
> 
> However unless you define what you want to do the numbers are useless.


Thank you very much for your answer. I would like to ask if there is any correlation between 78.4MB and 1.94GB*seconds. If so, how can it be calculated? CPU can only be calculated by how much time it takes, not how many Cs it consumes, right?

The 4.21 seconds of CPU time you mentioned is distributed throughout the cluster, of which 66.98 milliseconds are used for coordinator planning. Does it mean 12 C of 7 workers, that is, 84 C took a total of 4.21 seconds?

What I want is how many CPUs and GBs this SQL took in total; because the conventional calculation method for evaluating the resources of the entire cluster is that this SQL took 5C and 10G, and there are 10 SQLs of this type running simultaneously in the cluster, then I only need to ensure that there are 50C and 100G resources at the peak, which can ensure that this resource can be run to the maximum extent, and then there is basically no problem with the tasks at any other time point, so I need to use this to do resource evaluation and scale resources in real time.

**Reply by hashhar** [reply]
> I would like to ask if there is any correlation between 78.4MB and 1.94GB*seconds. If so, how can it be calculated?

There is a relation but indirect and hard to measure post-facto. Take the following example where metrics were sampled once a second and query took 10 seconds to finish.

```
1: 10MB
2: 12MB
3: 25MB
4: 74MB
5: 30MB
6: 50MB
7:20MB
8:10MB
9: 5MB
10: 1MB
```

In this example the "peak memory usage" metric is the max of all of the observed samples - 74MB at 4 seconds.
The "cumulative memory usage" is the technically the integral of memory usage at each instant in time over the 10 seconds. However since the memory usage cannot be observed as a continuous function and can only be sampled the integration is approximated numerically as AVG(usage) * duration. So in the above case it'd be 23.7MB (average) * 10 seconds = 237MB.

I hope the example clarifies the meaning of the metrics.

> The 4.21 seconds of CPU time you mentioned is distributed throughout the cluster, of which 66.98 milliseconds are used for coordinator planning. Does it mean 12 C of 7 workers, that is, 84 C took a total of 4.21 seconds?

CPU time is much simpler thankfully.

1 core provides 1 cpu-seconds of capacity per second. So 12C per worker each worker provides 12 cpu-seconds per second. With 7 workers that's 84 seconds of cpu time per second.

So in the metrics you share it means that the query used 4.21 cores for 1 full second (4.21 cpu-seconds) over the entire duration of the query. This is logically similar to the "cumulative memory usage" - i.e. total cpu consumed over the duration of the query. 4.21 cpu-seconds might have been consumed like:

```
1: 0s
2: 1s
3: 0s
4: 3.21s
```
or like:
```
1: 4s
2: 0.21s
```
or any other way. The metric only tells cumulative value.

> What I want is how many CPUs and GBs this SQL took in total; because the conventional calculation method for evaluating the resources of the entire cluster is that this SQL took 5C and 10G, and there are 10 SQLs of this type running simultaneously in the cluster, then I only need to ensure that there are 50C and 100G resources at the peak, which can ensure that this resource can be run to the maximum extent, and then there is basically no problem with the tasks at any other time point, so I need to use this to do resource evaluation and scale resources in real time.

in total is not a well-defined term. However your actual question (how large my cluster needs to be) is easier to answer.

Assuming n concurrent queries with exact same metrics as this query, you need:
total RAM across cluster: n * peak memory per query
total CPU across cluster: n * cpu-seconds per query

note that both above are the upper bound -  because it's possible not all queries will use the peak at the same point in time.

in the best case (where there is 0 concurrency or even with concurrency the peaks don't overlap) you need:
total RAM across cluster: peak memory per query + fixed headroom (let's assume 10% of peak memory for example, can be anything)
total CPU across cluster: doesn't matter, lack of cpu just makes things slow (indirectly it'll impact memory because if you have enough cpu to finish queries quickly they will run, use memory and then finish and release all memory quickly, with less cpu each query will potentially hold on to the memory for longer)

I really hope the above example shows you why this not a mathematical problem and you cannot use those metrics to make sizing decisions. In the worst case you'll provision so much resources that it'll be cost prohibitive (and utilization will be low) and on the other end of the spectrum you'll end up with cluster that keeps killing queries due to OOM or super slow performance due to cpu-time sharing.

The practical approach to sizing involves running the actual workload on the smallest possible cluster and keep adding resources until you meet your SLA requirements. That's the minimum resources you need. From there you can add more (or remove) based on how much variability you expect in your workload.

**Reply by hashhar** [reply]
Also note:

> resource evaluation and scale resources in real time

Doesn't help at all since the scheduler will only take into consideration resources available at the beginning of a query. If you add resources once a query starts running those will only be used for future queries. FTE mode is better at this but still adding resources in real-time and them being utilised in real-time doesn't happen (in any tool, not just Trino, even plain HTTP servers).

**Reply by A-little-bit-of-data** [reply]
Thank you very much for your answer. I think it's time for me to reconsider the issue of resource allocation.

**Comment by andrewa-habitat** [comment]
Is it possible to retrieve the `Resource Utilization Summary` as shown in the UI from a SQL query?

**Reply by A-little-bit-of-data** [reply]
> Is it possible to retrieve the `Resource Utilization Summary` as shown in the UI from a SQL query?

It's visible in the web UI, but when the worker runs Out of Memory (OOM) during their break time, it's no longer visible when you check back at work. The web UI can't save it for too long; I've tried saving it for a day, and it causes the coordinator OOM to restart.
